### PR TITLE
add noExAssign lint

### PR DESCRIPTION
**Pull request by weblancaster**
When using object assign cases I'm not really sure what path/solution to take, I'm guessing I would have to use recursion, I think the solution is around `AssignmentObjectPattern` and `AssignmentObjectPatternProperty` but no luck while trying to cover all scenarios.

Code exemplos:
```
"try { } catch (ex) { ({x: ex = 0} = {}); }",
"try { } catch (ex) { let a; ({x: a = ex = 0} = {}); }"
```

@sebmck please advise.

### CODE REVIEW

File: packages/@romejs/diagnostics/categories.ts [hunk 1]
```diff
@@ -11,99 +11,100 @@
 
 // made consistent.
 export type DiagnosticCategory =
```
Comment (weblancaster) [inline packages/@romejs/diagnostics/categories.ts]: seems like this file wasn't following the code format.

File: packages/@romejs/js-compiler/transforms/lint/noExAssign.test.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    "function foo() { try { } catch (e) { return false; } }"
+  ];
+
+  const invalidTestCases = [
+    "try { } catch (e) { e; e = 10; }",
+    "try { } catch (ex) { console.log('test'); ex = 10; }",
+    "try { } catch (ex) { [ex, test] = []; }",
+    "try { } catch ({message, name}) { message = 'test'; name = 10; }"
+    // "try { } catch (ex) { ({x: ex = 0} = {}); }",
+    // "try { } catch (ex) { let a; ({x: a = ex = 0} = {}); }"
```
Comment (weblancaster) [inline packages/@romejs/js-compiler/transforms/lint/noExAssign.test.ts]: I'm having issues with the two test cases above

**Comment by sebmck** [comment]
Thank you! Sorry for taking so long to review. I'm going to merge this as-is and will commit some of my nits over the top, such as using `forEach` instead of `for-of` etc. Feel free to put up another PR if you want to tackle those missing cases and I can try to provide some guidance there.

**Comment by weblancaster** [comment]
I feel bad now I made you work on this PR, sorry for that @sebmck

Thanks for taking care of this PR.

**Pull request merged**